### PR TITLE
TCES-19 Employers export data and filtering

### PR DESCRIPTION
**Pull request by tsids**
# Description

Implemented filtering for employers on the backend

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality

### CODE REVIEW

File: server/src/controllers/employer/getAllEmployers.js [hunk 1]
```diff
@@ -1,14 +1,60 @@
...
+
+    if (endDateAdded) {
+      query.creation_date = {
+        ...query.creation_date,
+        [Op.lte]: new Date(endDateAdded),
+      };
+    }
+
+    if (ownerId && ownerId !== "-1") {
+      query.creator = ownerId;
```
Comment (kevinle623) [inline server/src/controllers/employer/getAllEmployers.js]: can you change from `query.creator = ownerId` to `query.owner = ownerId`. mb for misleading, I fixed in my pr too

File: server/src/controllers/employer/getAllEmployers.js [hunk 2]
```diff
@@ -1,14 +1,60 @@
...
+      limit: pageSize,
+      offset: offsetSize,
+    });
+
+    const totalEmployers = await Employer.count({ where: query });
 
     return res.status(200).json({
       status: "success",
       message: "All employers found successfully",
       data: employers,
```
Comment (kevinle623) [inline server/src/controllers/employer/getAllEmployers.js:71]: looks good, I was just wondering if you can add the following attributes to the response body as well:

can you fetch all the unique owner ids in the table similar to so:
```
const uniqueOwners = await Employer.findAll({
      attributes: [[Sequelize.fn("DISTINCT", Sequelize.col("owner")), "owner"]],
      raw: true,
    });
```

After then attach it to the response body, something like

```
return res.status(200).json({
      status: "success",
      message: "All employers found successfully",
      data: employers,
      total: totalEmployers,
      uniqueOwners: uniqueOwners.map((owner) => owner.owner),
    });

```

**Review by kevinle623 — COMMENTED** [review]
looks good, left a few comments when you have time to check over, and for the failing tests it might be a mocking issue, I had to do the same in [my pr](https://github.com/uoftblueprint/tces/pull/41) if you want to refer to it as well

**Comment by tsids** [comment]
@kevinle623
I'm not sure how to fix the failed server check. I didn't change anything substantial in the tests file.

**Comment by kevinle623** [comment]
> @kevinle623 I'm not sure how to fix the failed server check. I didn't change anything substantial in the tests file.

Hey Taha, no worries, there was just some types and variables that weren't imported at the top of the file and some typos. For example, Sequalize needed to be declared at the top to prevent a type error. I have just pushed some changes when you have time to fetch/pull to check out. But it should be good now. One of the main changes was adding the following to prevent uniqueOwners type error in the vitests. 

```
const uniqueOwners = await JobLead.findAll({
      attributes: [[Sequelize.fn("DISTINCT", Sequelize.col("owner")), "owner"]],
      raw: true,
    });
    const uniqueOwnersList = Array.isArray(uniqueOwners)
      ? uniqueOwners.map((owner) => owner.owner)
      : [];
```

tests should be passing now. :)

**Review by kevinle623 — DISMISSED** [review]

**Comment by tsids** [comment]
> Hey Taha, no worries, there was just some types and variables that weren't imported at the top of the file and some typos. For example, Sequalize needed to be declared at the top to prevent a type error. I have just pushed some changes when you have time to fetch/pull to check out. But it should be good now. One of the main changes was adding the following to prevent uniqueOwners type error in the vitests.
> 
> ```
> const uniqueOwners = await JobLead.findAll({
>       attributes: [[Sequelize.fn("DISTINCT", Sequelize.col("owner")), "owner"]],
>       raw: true,
>     });
>     const uniqueOwnersList = Array.isArray(uniqueOwners)
>       ? uniqueOwners.map((owner) => owner.owner)
>       : [];
> ```
> 
> tests should be passing now. :)

Thanks Kevin, I appreciate it.

**Review by kevinle623 — DISMISSED** [review]

File: server/src/controllers/employer/getAllEmployers.js [hunk 3]
```diff
@@ -1,14 +1,76 @@
...
+    if (employerName) {
+      query.name = { [Op.like]: `%${employerName}%` };
+    }
+    if (phoneNumber) {
+      query.phone_number = { [Op.like]: `%${phoneNumber}%` };
+    }
+    if (startDateAdded) {
+      const startDate = new Date(startDateAdded);
+      startDate.setHours(0, 0, 0, 0);
+      query.date_added = { [Op.gte]: new Date(startDateAdded) };
```
Comment (DanielDervishi) [inline server/src/controllers/employer/getAllEmployers.js]: Change this to use startDate

File: server/src/controllers/employer/getAllEmployers.js [hunk 4]
```diff
@@ -1,14 +1,76 @@
...
+    }
+
+    if (endDateAdded) {
+      const endDate = new Date(endDateAdded);
+      endDate.setHours(23, 59, 59, 999);
+      query.date_added = {
+        ...query.date_added,
+        [Op.lte]: new Date(endDateAdded),
```
Comment (DanielDervishi) [inline server/src/controllers/employer/getAllEmployers.js]: change this to use endDate

**Review by DanielDervishi — COMMENTED** [review]
Great work Taha! Just left two quick comments, but it should be ready to be merged after this.

**Review by kevinle623 — DISMISSED** [review]

**Review by kevinle623 — APPROVED** [review]

**Review by DanielDervishi — APPROVED** [review]
Awesome work!

**Pull request merged**